src/analysis/calculus/local_extr.lean

```diff
@@ -0,0 +1,293 @@
+/-
+Copyright (c) 2019 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+
+import topology.local_extr analysis.calculus.deriv
+
+/-! # Local extrema of smooth functions
+
+## Main definitions
+
+In a real normed space `E` we define `pos_tangent_cone_at (s : set E) (x : E)`.
+This would be the same as `tangent_cone_at ℝ≥0 s x` if we had a theory of normed semifields.
+This set is used in the proof of Fermat's Theorem (see below), and can be used to formalize
+[Lagrange multipliers](https://en.wikipedia.org/wiki/Lagrange_multiplier) and/or
+[Karush–Kuhn–Tucker conditions](https://en.wikipedia.org/wiki/Karush–Kuhn–Tucker_conditions).
+
+## Main statements
+
+First we prove that `0 ≤ f' y` whenever `a` is a local maximum of `f` on `s`,
+`f` has derivative `f'` at `a` within `s`, and `y` belongs to the positive tangent cone
+of `s` at `a`. Hence if both `y` and `-y` belong to the positive tangent cone, then `f' y = 0`.
+These facts are used to prove Fermat's Theorem: the derivative of a differentiable function
+at a local extremum point equals zero.
```

Comment:
Could you give the names of the formalized theorems, to help the reader?

Reply:
Added `References`. Did I miss any named theorem? Then probably I'm not aware that it has a name.

Reply:
Sorry, I wasn't clear. I mean the Lean names, like `is_local_max_on.has_fderiv_within_at_nonpos` or `is_local_extr.has_fderiv_at_eq_zero` (or versions with `*`). You don't need to give all of them, it is just to give some pointers to the reader.